codec/api.go

```diff
@@ -1,48 +1,124 @@
+// but aren't promising memory reuse if kept around.
 //
-// By convention, codec packages that have a multicodec contract will also have
-// a package-scope exported function called Encode which also matches this interface,
-// and is the equivalent of creating a zero-value ReusableEncoder (aka, default config)
-// and using its Encode method.
-// This package-scope function will typically also internally use a sync.Pool
+// By convention, a codec package that expects to fulfill a multicodec contract will also have
+// a package-scope exported function called Encode or Decode which also matches this interface,
+// and is the equivalent of creating a zero-value ReusableEncoder or ReusableDecoder (aka, default config)
+// and using its Encode or Decode methods.
+// This package-scope function may also internally use a sync.Pool
 // to keep some ReusableEncoder values on hand to avoid unnecesary allocations.
 //
-// Note that a ReusableEncoder type that supports configuration options
+// Note that an EncoderConfig or DecoderConfig type that supports configuration options
 // does not functionally expose those options when invoked by the multicodec system --
 // multicodec indicators do not provide room for extended configuration info.
 // Codecs that expose configuration options are doing so for library users to enjoy;
 // it does not mean those non-default configurations will necessarly be available
 // in all scenarios that use codecs indirectly.
 // There is also no standard interface for such configurations: by nature,
-// if they exist at all, they vary per codec.
-type Encoder = ipld.Encoder
+// if they exist at all, they tend to vary per codec.
+type (
+	// Encoder defines the shape of a function which traverses a Node tree
+	// and emits its data in a serialized form into an io.Writer.
+	//
+	// The dual of Encoder is a Decoder, which takes a NodeAssembler
+	// and fills it with deserialized data consumed from an io.Reader.
+	// Typically, Decoder and Encoder functions will be found in pairs,
+	// and will be expected to be able to round-trip each other's data.
+	//
+	// Encoder functions can be used directly.
+	// Encoder functions are also often used via a LinkSystem when working with content-addressed storage.
+	// LinkSystem methods will helpfully handle the entire process of traversing a Node tree,
+	// encoding this data, hashing it, streaming it to the writer, and committing it -- all as one step.
+	//
+	// An Encoder works with Nodes.
+	// If you have a native golang structure, and want to serialize it using an Encoder,
+	// you'll need to figure out how to transform that golang structure into an ipld.Node tree first.
+	//
+	// It may be useful to understand "multicodecs" when working with Encoders.
+	// In IPLD, a system called "multicodecs" is typically used to describe encoding foramts.
+	// A "multicodec indicator" is a number which describes an encoding;
```

Comment:
"multicodec code" is the terminology we've been mostly standardising on recently

Reply:
Updated to "multicodec indicator code".

Which is verbose, but extremely accurate.  (And tbh I can't stand the stutter of "codec code" otherwise.)